File: Sources/UI/LifetimeCard.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import SwiftUI

/// **All time, in History between the trend deck and the sessions** (Nuri, 2026-09-20).
/// Sessions, pulls, time under tension, volume, days trained, climbing days, heaviest
/// pull and the start date. No chart or trend — the cards above own those; this is the
/// odometer, sitting over the sessions it adds up.
///
/// A LEDGER, not tiles: a three-column grid read as ragged ("49" left a hole, "4 hr,
/// 40 min" crowded the edge). Label-left, value-right rows like the Device card put every
/// value on one right-hand axis.
///
/// Folded from denormalized columns (`Collection.lifetime`) off History's own query: a
/// row per session, never a decode.
struct LifetimeCard: View {
    var stats: LifetimeStats
    @Environment(\.weightUnit) private var weightUnit

    var body: some View {
        MaterialCard(surface: .flat) {
            VStack(alignment: .leading, spacing: 12) {
                CapsLabel(title)
                if stats.isEmpty {
                    Text("Your all-time numbers land here after the first session.")
                        .foregroundStyle(Ink.secondary)
                        .fixedSize(horizontal: false, vertical: true)
                } else {
                    row(String(localized: "Sessions"), stats.sessions.formatted())
                    row(String(localized: "Pulls"), stats.pulls.formatted())
                    row(String(localized: "Under tension"), heldText)
                    row(String(localized: "Volume"), volumeText)
                    row(String(localized: "Days trained"), stats.daysTrained.formatted())
                    if stats.climbDays > 0 {
                        row(String(localized: "Climbing days"), stats.climbDays.formatted())
                    }
                    row(String(localized: "Heaviest pull"), weightUnit.text(stats.heaviestPullKg))
                    Text("Volume is load × pulls, added up. Under tension is every second on the edge.")
                        .font(.system(.caption))
                        .foregroundStyle(Ink.secondary)
                        .fixedSize(horizontal: false, vertical: true)
                        .padding(.top, 2)
                }
            }
            .font(.system(.subheadline))
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel(spoken)
        .accessibilityIdentifier("settings.lifetime")
    }

    private var title: String {
        guard let since = stats.since else { return String(localized: "All time") }
        let from = since.date().formatted(.dateTime.day().month(.abbreviated).year())
        return String(localized: "All time · since \(from)")
    }

    /// The Device card's row, value in primary ink with a little weight: facts about the
    /// person, not the link's status.
    private func row(_ label: String, _ value: String) -> some View {
        LabeledContent(label) {
            Text(value)
                .fontWeight(.semibold)
                .monospacedDigit()
                .foregroundStyle(Ink.primary)
        }
    }

    /// Two units, never three: "4 hr, 40 min" reads; add seconds and it is a stopwatch.
    private var heldText: String {
        Duration.seconds(stats.heldSeconds)
            .formatted(.units(allowed: [.hours, .minutes, .seconds], width: .abbreviated,
                              maximumUnitCount: 2))
    }

    /// Tonnes past a thousand kilograms ("20.3 t" is an odometer, "20 300 kg" a spreadsheet).
    /// Pounds stay pounds: a ton is two different weights in English.
    private var volumeText: String {
        switch weightUnit {
        case .kg:
            if stats.volumeKg >= 1000 {
                return String(localized: "\((stats.volumeKg / 1000).formatted(.number.precision(.fractionLength(1)))) t")
            }
            return "\(stats.volumeKg.formatted(.number.precision(.fractionLength(0)))) \(weightUnit.symbol)"
        case .lb:
            return "\(weightUnit.fromKg(stats.volumeKg).formatted(.number.precision(.fractionLength(0)))) \(weightUnit.symbol)"
        }
    }

    private var spoken: String {
        guard !stats.isEmpty else { return String(localized: "All time. Your numbers land here after the first session.") }
        var parts = [
            String(localized: "\(stats.sessions) sessions"),
            String(localized: "\(stats.pulls) pulls"),
            String(localized: "\(heldText) under tension"),
            String(localized: "volume \(volumeText)"),
            String(localized: "\(stats.daysTrained) days trained"),
        ]
        if stats.climbDays > 0 { parts.append(String(localized: "\(stats.climbDays) climbing days")) }
        parts.append(String(localized: "heaviest pull \(weightUnit.number(stats.heaviestPullKg)) \(weightUnit.spokenName)"))
        return "\(title). \(parts.joined(separator: ", "))."
    }
}

File: android/app/src/main/kotlin/run/nuri/getagrip/ui/history/LifetimeCard.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ui.history

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import run.nuri.getagrip.data.LifetimeStats
import run.nuri.getagrip.ui.components.CapsLabel
import run.nuri.getagrip.ui.l10n.tr
import run.nuri.getagrip.ui.theme.LocalGripPalette
import run.nuri.getagrip.ui.theme.Metrics
import run.nuri.getagrip.ui.units.WeightUnit
import run.nuri.getagrip.ui.units.WeightUnits
import java.text.NumberFormat
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle

/// **All time, in History between the calendar and the sessions** (Nuri, 2026-09-20:
/// "lifetime stats — number of routines done, total load lifetime, anything else?").
/// Sessions, pulls, time under tension, volume, days trained, climbing days, the heaviest
/// pull, and the date they count from. Nothing here is a chart or a trend — the calendar
/// above it owns those — this is the odometer, and it sits right over the sessions it adds
/// up.
///
/// A LEDGER, not tiles: label left, value right, the rows the Settings device card is drawn
/// as. The iOS first cut was a three-column grid and read as ragged — a "49" left a hole two
/// numbers wide (Nuri: "spacing on this is kind of ugly"). Folded from denormalized columns
/// (`Collection<WorkoutLogEntity>.lifetime`), so it costs a row per session, never a decode.
@Composable
fun LifetimeCard(stats: LifetimeStats, modifier: Modifier = Modifier) {
    val palette = LocalGripPalette.current
    val title = stats.since?.let { since ->
        tr("All time · since %s", since.localDate().format(DateTimeFormatter.ofLocalizedDate(FormatStyle.MEDIUM)))
    } ?: tr("All time")
    Surface(
        shape = RoundedCornerShape(Metrics.radiusCard),
        color = palette.card,
        modifier = modifier.fillMaxWidth().semantics(mergeDescendants = true) {},
    ) {
        Column(Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
            CapsLabel(title)
            if (stats.isEmpty) {
                Text(
                    tr("Your all-time totals appear after your first session."),
                    style = MaterialTheme.typography.bodyMedium,
                    color = palette.inkSecondary,
                )
            } else {
                val integers = NumberFormat.getIntegerInstance()
                LedgerRow(tr("Sessions"), integers.format(stats.sessions))
                LedgerRow(tr("Pulls"), integers.format(stats.pulls))
                LedgerRow(tr("Under tension"), heldText(stats.heldSeconds))
                LedgerRow(tr("Volume"), volumeText(stats.volumeKg))
                LedgerRow(tr("Days trained"), integers.format(stats.daysTrained))
                if (stats.climbDays > 0) LedgerRow(tr("Climbing days"), integers.format(stats.climbDays))
                LedgerRow(tr("Heaviest pull"), WeightUnits.current.text(stats.heaviestPullKg))
                Text(
                    tr("Volume is load × pulls. Under tension is total time on the edge."),
                    style = MaterialTheme.typography.bodySmall,
                    color = palette.inkTertiary,
                    modifier = Modifier.padding(top = 2.dp),
                )
            }
        }
    }
}

/// The Settings device card's row, with the value in primary ink and a little weight: these
/// are facts about the person, not the link's status.
@Composable
private fun LedgerRow(label: String, value: String) {
    val palette = LocalGripPalette.current
    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
        Text(label, style = MaterialTheme.typography.bodyMedium, color = palette.inkSecondary)
        Text(
            value,
            style = MaterialTheme.typography.bodyMedium.copy(fontFeatureSettings = "tnum"),
            fontWeight = FontWeight.SemiBold,
            color = palette.inkPrimary,
        )
    }
}

/// Hours and minutes once there are hours, minutes before — two units, never three.
@Composable
private fun heldText(seconds: Double): String {
    val total = seconds.toLong()
    val hours = (total / 3600).toInt()
    val minutes = ((total % 3600) / 60).toInt()
    return if (hours > 0) tr("%d hours and %d minutes", hours, minutes) else tr("%d min", minutes)
}

/// Tonnes once kilograms pass a thousand — "20.3 t" is the odometer reading, "20 300 kg" is
/// a spreadsheet. Pounds stay pounds: a ton is two different weights in English.
@Composable
private fun volumeText(volumeKg: Double): String {
    val unit = WeightUnits.current
    val oneDecimal = NumberFormat.getNumberInstance().apply { maximumFractionDigits = 1; minimumFractionDigits = 1 }
    val whole = NumberFormat.getIntegerInstance()
    return when (unit) {
        WeightUnit.kg ->
            if (volumeKg >= 1000) tr("%s t", oneDecimal.format(volumeKg / 1000))
            else "${whole.format(volumeKg)} ${unit.symbol}"
        WeightUnit.lb -> "${whole.format(unit.fromKg(volumeKg))} ${unit.symbol}"
    }
}
